### PR TITLE
feat: list objects pagination

### PR DESCRIPTION
**Pull request by BarryTong65**
### Description

Create pagination feature for ListObjectByBucketName.

### Rationale

Avoid returning too much data at once by implementing pagination feature

### Example

N/A

### Changes

Notable changes: 
* pagination feature for Listobjectbybucketname
* add get user buckets default limit

### CODE REVIEW

File: proto/service/metadata/types/metadata.proto [hunk 1]
```diff
@@ -65,6 +65,10 @@ message ListObjectsByBucketNameRequest {
   string bucket_name = 1;
   // account_id is the account address of user
   string account_id = 2;
+  // max_keys sets the maximum number of keys returned in the response
+  uint64  max_keys = 3;
```
Comment (sysvm) [inline proto/service/metadata/types/metadata.proto:69]: Maybe use start, limit, offset to do pagination is better?

File: service/metadata/service/object.go [hunk 2]
```diff
@@ -3,21 +3,34 @@ package service
 import (
 	"context"
 
+	"cosmossdk.io/math"
 	model "github.com/bnb-chain/greenfield-storage-provider/store/bsdb"
```
[the remark this answers is not in the document]
Comment (BarryTong65) [reply service/metadata/service/object.go:7]: image
"cosmossdk.io/math" belongs to a third-party library, not std.

As you mentioned before, imported pkg shoud be grouped in std, third-party, inner pkg

[the remark this answers is not in the document]
Comment (sysvm) [reply service/metadata/service/object.go:7]: I mean `model "github.com/bnb-chain/greenfield-storage-provider/store/bsdb"` should be sorted.

File: proto/service/metadata/types/metadata.proto (hunk 1 above)
Comment (sysvm) [reply proto/service/metadata/types/metadata.proto:69]: OK, thanks!

File: service/metadata/service/object.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (BarryTong65) [reply service/metadata/service/object.go]: sorry about that, fixed now

**Review by annielz — APPROVED** [review]

File: service/gateway/metadata_handler.go [hunk 3]
```diff
@@ -102,8 +105,32 @@ func (gateway *Gateway) listObjectsByBucketNameHandler(w http.ResponseWriter, r
...
 
+	if reqContext.maxKeys != "" {
+		maxKeys, err = strconv.ParseUint(reqContext.maxKeys, 10, 64)
+		if err != nil {
+			log.Errorw("failed to parse maxKeys", "max_keys", reqContext.maxKeys, "error", err)
+			errDescription = InvalidMaxKeys
+			return
+		}
+		// maxKeys should > 0
+		if maxKeys <= 0 {
```
Comment (will-2012) [inline service/gateway/metadata_handler.go:116]: uint64 will not be less than 0

File: service/gateway/metadata_handler.go [hunk 4]
```diff
@@ -102,8 +105,32 @@ func (gateway *Gateway) listObjectsByBucketNameHandler(w http.ResponseWriter, r
 		return
 	}
 
+	if reqContext.maxKeys != "" {
+		maxKeys, err = strconv.ParseUint(reqContext.maxKeys, 10, 64)
```
Comment (will-2012) [inline service/gateway/metadata_handler.go:109]: reuse util.StringToUint64?

File: service/gateway/metadata_handler.go [hunk 5]
```diff
@@ -102,8 +105,32 @@ func (gateway *Gateway) listObjectsByBucketNameHandler(w http.ResponseWriter, r
...
+		// maxKeys should > 0
+		if maxKeys <= 0 {
+			log.Errorw("failed to check maxKeys", "max_keys", reqContext.maxKeys, "error", err)
+			errDescription = InvalidMaxKeys
+			return
+		}
+	}
+
+	// startAfter is an optional input, we only check its format when user input the value
+	if ok = IsHexHash(reqContext.startAfter); !ok && reqContext.startAfter != "" {
```
Comment (will-2012) [inline service/gateway/metadata_handler.go:124]: What is the meaning of `startAfter`?

File: service/gateway/request_util.go [hunk 6]
```diff
@@ -75,6 +77,8 @@ func newRequestContext(r *http.Request) *requestContext {
 		bucketName: vars["bucket"],
 		objectName: vars["object"],
 		accountID:  vars["account_id"],
+		maxKeys:    r.URL.Query().Get("max_keys"),
+		startAfter: r.URL.Query().Get("start_after"),
```
Comment (will-2012) [inline service/gateway/request_util.go:81]: The parts that are not common can be parsed in the corresponding handler.

File: store/bsdb/object.go [hunk 7]
```diff
@@ -1,16 +1,34 @@
...
+			Limit(limit).
+			Find(&objects).Error
+		return objects, err
+	}
+	// select objects after a specific key.
 	err = b.db.Table((&Object{}).TableName()).
 		Select("*").
-		Where("bucket_name = ?", bucketName).
-		Order("create_at desc").
+		Where("bucket_name = ? and object_id <= ?", bucketName, startAfter).
```
[the remark this answers is not in the document]
Comment (BarryTong65) [reply store/bsdb/object.go:29]: Yes, the user can use startAfter by passing in an object ID to paginate through objects that come after the specified key. By default, the data retrieved will be the latest objects

File: service/gateway/metadata_handler.go (hunk 5 above)
Comment (BarryTong65) [reply service/gateway/metadata_handler.go:124]: Compared to the traditional limit and offset, startAfter is the specified object ID, and maxKeys is the number of objects requested by the user

File: store/bsdb/object.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (will-2012) [reply store/bsdb/object.go:29]: In the AWS s3, start_after is have clear semantics. And here We have different semantics, is it better to change the name?

File: service/gateway/request_util.go (hunk 6 above)
Comment (BarryTong65) [reply service/gateway/request_util.go]: fixed

File: service/gateway/metadata_handler.go (hunk 4 above)
Comment (BarryTong65) [reply service/gateway/metadata_handler.go]: fixed

File: service/gateway/metadata_handler.go (hunk 3 above)
Comment (BarryTong65) [reply service/gateway/metadata_handler.go]: fixed

File: service/metadata/service/object.go [hunk 8]
```diff
@@ -3,21 +3,34 @@ package service
...
+		objects               []*model.Object
+		keyCount              uint64
+		isTruncated           bool
+		nextContinuationToken string
+		maxKeys               uint64
+	)
+
+	maxKeys = req.MaxKeys
+	// if the user does not provide any input parameters, default values will be used
+	if req.MaxKeys == 0 {
```
[the remark this answers is not in the document]
Comment (BarryTong65) [reply service/metadata/service/object.go:28]: fixed, additionally add ListObjectsLimitSize(1000) if user input maxKyes > 1000.
ListObjectsDefaultMaxKeys is the default value used when the user does not specify a value for MaxKeys, while ListObjectsLimitSize limits the number of results returned to a maximum of 1000 when the user inputs a value greater than 1000

**Pull request closed without merging**